### PR TITLE
Allow using Azure rm legacy hostnames

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This allows the azure_rm inventory plugin to return the same hostnames as `azure_rm.py`

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
azure_rm inventory plugin

##### ADDITIONAL INFORMATION
Tested `ansible-inventory` calls between plugin and inventory script side-by-side.

This needs an additional fix due to this error:

```
 [WARNING]:  * Failed to parse /Users/alancoding/Documents/repos/ansible-inventory-file-examples/private/azure/azure_rm.yml with auto plugin: can't add group to itself

  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/manager.py", line 272, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/auto.py", line 58, in parse
    plugin.parse(inventory, loader, path, cache=cache)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/azure_rm.py", line 263, in parse
    self._get_hosts()
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/azure_rm.py", line 344, in _get_hosts
    self._add_host_to_keyed_groups(constructable_config_keyed_groups, h.hostvars, inventory_hostname, strict=constructable_config_strict)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/plugins/inventory/__init__.py", line 431, in _add_host_to_keyed_groups
    self.inventory.add_child(result_gname, host)
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/data.py", line 262, in add_child
    g.add_child_group(self.groups[child])
  File "/Users/alancoding/Documents/repos/ansible/lib/ansible/inventory/group.py", line 177, in add_child_group
    raise Exception("can't add group to itself")
```

This is encountered in the cases where groupings are used of:

 - name
 - security group
 - resource group

Either people will have to lose those groups (unacceptable), or another fix will be needed to the general constructible plugin to allow non-strict recursion. Ping @wenottingham @s-hertel

### CODE REVIEW

File: lib/ansible/plugins/inventory/azure_rm.py [hunk 1]
```diff
@@ -74,6 +74,13 @@
           type: bool
           default: False
           version_added: '2.8'
+        use_contrib_script_host_names:
```
Comment (bcoca) [inline lib/ansible/plugins/inventory/azure_rm.py]: i would name it to 'plain_host_names' and add information about these not being unique and that when  `False` and additional 'hashing' is added to the name to ensure uniqueness.

Comment (AlanCoding) [reply lib/ansible/plugins/inventory/azure_rm.py]: change made

Comment (kdelee) [reply lib/ansible/plugins/inventory/azure_rm.py]: this has been done

**Review by nitzmahone — APPROVED** [review]

**Comment by abadger** [comment]
@AlanCoding Can you add a changelog for this and then ping me to merge it?

**Comment by nitzmahone** [comment]
@AlanCoding should probably be under a `minor_changes` label

**Comment by abadger** [comment]
rebuild_merge

**Comment by rolandjohann** [comment]
currently we are running in similar issues caused by non predictable `inventory_hostname`s. Wouldn't it be better to be able to configure the hostname, like it is possible at `aws_ec2` inventory plugin or just by interpret jinja2 template strings based on host metadata returned by azure API? 

I recognized this issue because I wanted to implement `inventory_hostname` configuration and stumbled over this new feature.

**Comment by AlanCoding** [comment]
If you want both unique host names and the ability to reliably identify hosts, you could use keyed_groups to create a group based on the specific hostvar that you want to use for identification. The group name will still work as a host pattern.

**Comment by rolandjohann** [comment]
@AlanCoding thanks for quick answer!

The grouping came to mind, but seems a bit hacky to achieve known behavior of `inventory_hostname` 😉

There are two issues:
* you must define `ansible_host` to be able to connect to the instance (which can/should be moved to `inventory_hostname` definition)
* you aren't able to target the specific host via `ansible-playbook --limit "<inventory_hostname>" ...`
  * this can be worked around by creating separate groups for each host, but counterintuitive

We can get both features by something like
```py
constructable_config_inventory_hostname_expression = self.get_option('inventory_hostname_expression')

for h in self._hosts:
    inventory_hostname = self._get_hostname(h, constructable_config_inventory_hostname_expression)

def _get_hostname(self, host, template):
    if template:
        return self._compose(template, host.hostvars)
    else:
        # FUTURE: configurable hostname sources
        return host.default_inventory_hostname
```
and configure plugin to use the `name` to achieve legacy naming
```
inventory_hostname_expression: name
```

Doing this we are also able to define whatever value we want:
```
inventory_hostname_expression: "tags.fqdn"
inventory_hostname_expression: "private_ipv4_addresses + public_dns_hostnames + [name] | first"
```

**Pull request merged**